### PR TITLE
#85 - IntegerAnonymizer - add possibility to anonymize value by adding noise

### PR DESCRIPTION
**Pull request by SimonMellerin**

### CODE REVIEW

File: src/Anonymization/Anonymizer/Core/IntegerAnonymizer.php [hunk 1]
```diff
@@ -23,18 +26,67 @@ class IntegerAnonymizer extends AbstractAnonymizer
...
+        if ($this->options->has('min') && $this->options->has('max')) {
+            $this->anonymizeWithMinAndMax(
+                $update,
+                (int) $this->options->get('min'),
+                (int) $this->options->get('max')
+            );
+        } elseif ($this->options->has('delta')) {
+            $this->anonymizeWithDelta($update, (int) $this->options->get('delta'));
+        } elseif ($this->options->has('percent')) {
+            $this->anonymizeWithPercent($update, (int) $this->options->get('percent'));
```
Comment (Lonnytunes) [inline src/Anonymization/Anonymizer/Core/IntegerAnonymizer.php:39]: Shouldn't we ensure that `delta` and `percent` options contain a positive integer ?

Or maybe we could just add `$delta = \abs($delta)` before using it in `anonymizeWithDelta()`? The same thing for the percent version.

File: src/Anonymization/Anonymizer/Core/IntegerAnonymizer.php [hunk 2]
```diff
@@ -23,18 +26,67 @@ class IntegerAnonymizer extends AbstractAnonymizer
...
+
+    private function anonymizeWithPercent(Update $update, int $percent): void
+    {
+        $expr = $update->expression();
+
+        $update->set(
+            $this->columnName,
+            $expr->cast(
+                $expr->raw(
+                    '? * (?) / 100',
```
Comment (Lonnytunes) [inline src/Anonymization/Anonymizer/Core/IntegerAnonymizer.php:79]: I wonder about the parentheses. Are they really useful?

Comment (SimonMellerin) [reply src/Anonymization/Anonymizer/Core/IntegerAnonymizer.php:79]: yep, it does not work without them

File: src/Anonymization/Anonymizer/Core/IntegerAnonymizer.php (hunk 1 above)
Comment (SimonMellerin) [reply src/Anonymization/Anonymizer/Core/IntegerAnonymizer.php:39]: I think we should trigger an exception if percent or delta are negative. Putting such values for those options will be (most of the time) the result of a mistake or a misunderstood on how this anonymizer works.

File: src/Anonymization/Anonymizer/Core/IntegerAnonymizer.php (hunk 2 above)
Comment (Lonnytunes) [reply src/Anonymization/Anonymizer/Core/IntegerAnonymizer.php:79]: Quite strange :thinking:

Comment (SimonMellerin) [reply src/Anonymization/Anonymizer/Core/IntegerAnonymizer.php:90]: Well i guess it depends of what the `$this->getRandomIntExpression` returns (in SQL). 

It only failed for Mysql I think.

Comment (SimonMellerin) [reply src/Anonymization/Anonymizer/Core/IntegerAnonymizer.php:90]: No I think it was SQLite :

```php

        return $this->formatRaw(
            new Raw(
                'abs(random()) % (? - ?) + ?',
                [$max, $min, $min]
            ),
            $context,
        );
```
(from the query builder source code)

Comment (Lonnytunes) [reply src/Anonymization/Anonymizer/Core/IntegerAnonymizer.php:90]: :+1:

Comment (Lonnytunes) [reply src/Anonymization/Anonymizer/Core/IntegerAnonymizer.php:90]: Finally, shouldn't we fix this method from the `SqliteWriter` instead?

Comment (SimonMellerin) [reply src/Anonymization/Anonymizer/Core/IntegerAnonymizer.php:90]: see https://github.com/makinacorpus/php-query-builder/issues/6

File: src/Anonymization/Anonymizer/Core/IntegerAnonymizer.php (hunk 1 above)
Comment (Lonnytunes) [reply src/Anonymization/Anonymizer/Core/IntegerAnonymizer.php]: I agree with this solution :ok_hand:

File: src/Anonymization/Anonymizer/Core/IntegerAnonymizer.php (hunk 2 above)
Comment (Lonnytunes) [reply src/Anonymization/Anonymizer/Core/IntegerAnonymizer.php:90]: > I agree but as it comes from a dependency, I suggest we keep it like this for now

Very true :sweat_smile:

**Review by Lonnytunes — APPROVED** [review]

**Pull request merged**